File: tests/test_adapter.py
from __future__ import annotations

import json
from pathlib import Path

import pytest

from hp_guard import (
    AdapterError,
    AuditError,
    AuditLog,
    AuditedPolicyStore,
    EnforcementRequest,
    InlineEnforcementAdapter,
    PolicyCall,
)


FIXTURE_PATH = Path(__file__).parents[1] / "conformance" / "cases" / "inline_adapter_v1.json"


def _fixture() -> dict:
    return json.loads(FIXTURE_PATH.read_text())


def _records(path: Path) -> list[dict]:
    return [json.loads(line) for line in path.read_text().splitlines()]


def _adapter(tmp_path, *, now_unix_ms: int = 1000):
    fixture = _fixture()
    audit_path = tmp_path / "audit.jsonl"
    store = AuditedPolicyStore(fixture["policy"], AuditLog(audit_path))
    return InlineEnforcementAdapter(store, now_unix_ms=lambda: now_unix_ms), audit_path


def test_adapter_matches_shared_allow_and_non_allow_responses(tmp_path):
    fixture = _fixture()
    adapter, _ = _adapter(tmp_path, now_unix_ms=fixture["now_unix_ms"])

    responses = [
        adapter.authorize(EnforcementRequest.from_dict(request)).to_dict()
        for request in fixture["requests"]
    ]

    assert responses == fixture["responses"]


@pytest.mark.parametrize("case", _fixture()["error_cases"], ids=lambda case: case["name"])
def test_adapter_rejects_shared_errors_without_an_authorization_record(tmp_path, case):
    fixture = _fixture()
    adapter, audit_path = _adapter(tmp_path, now_unix_ms=fixture["now_unix_ms"])
    before = _records(audit_path)

    if case["error"] == "invalid_request":
        with pytest.raises(AdapterError) as raised:
            EnforcementRequest.from_dict(case["request"])
    else:
        with pytest.raises(AdapterError) as raised:
            adapter.authorize(EnforcementRequest.from_dict(case["request"]))

    assert raised.value.code == case["error"]
    assert _records(audit_path) == before


@pytest.mark.parametrize(
    "payload",
    [
        {"caller_id": "caller", "correlation_id": "id", "deadline_unix_ms": True, "call": {}},
        {"caller_id": "caller", "correlation_id": "id", "deadline_unix_ms": -1, "call": {}},
        {"caller_id": "caller", "correlation_id": "id", "deadline_unix_ms": 1, "call": {}},
        {
            "caller_id": "caller",
            "correlation_id": "id",
            "deadline_unix_ms": 1,
            "call": {"agent": None, "tool": None, "args": [1], "user": None, "context": {}},
        },
        {
            "caller_id": "caller",
            "correlation_id": "id",
            "deadline_unix_ms": 1,
            "call": {"agent": None, "tool": None, "args": [], "user": None, "context": {"key": 1}},
        },
        {"caller_id": "x" * 257, "correlation_id": "id", "deadline_unix_ms": 1, "call": {}},
        {"caller_id": "é" * 129, "correlation_id": "id", "deadline_unix_ms": 1, "call": {}},
        {"caller_id": "caller", "correlation_id": "x" * 257, "deadline_unix_ms": 1, "call": {}},
    ],
)
def test_request_boundary_strictly_validates_the_complete_request(payload):
    with pytest.raises(AdapterError) as raised:
        EnforcementRequest.from_dict(payload)

    assert raised.value.code == "invalid_request"


def test_request_constructor_cannot_bypass_validation_with_a_malformed_call():
    with pytest.raises(AdapterError) as raised:
        EnforcementRequest("caller", "id", 1, PolicyCall(args=("not", "a", "list")))

    assert raised.value.code == "invalid_request"


@pytest.mark.parametrize(
    "call",
    [
        {"agent": "\ud800", "tool": "shell", "args": [], "user": None, "context": {}},
        {"agent": None, "tool": "\ud800", "args": [], "user": None, "context": {}},
        {"agent": None, "tool": "shell", "args": [], "user": "\ud800", "context": {}},
        {"agent": None, "tool": "shell", "args": ["\ud800"], "user": None, "context": {}},
        {"agent": None, "tool": "shell", "args": [], "user": None, "context": {"\ud800": "value"}},
        {"agent": None, "tool": "shell", "args": [], "user": None, "context": {"key": "\ud800"}},
    ],
)
def test_request_boundary_rejects_lone_surrogates_in_all_call_strings(call):
    with pytest.raises(AdapterError) as raised:
        EnforcementRequest.from_dict(
            {"caller_id": "caller", "correlation_id": "id", "deadline_unix_ms": 1, "call": call}
        )

    assert raised.value.code == "invalid_request"


@pytest.mark.parametrize(
    "request_factory",
    [
        lambda: EnforcementRequest.from_dict(
            {
                "caller_id": "caller",
                "correlation_id": "id",
                "deadline_unix_ms": 1 << 64,
                "call": {"agent": None, "tool": None, "args": [], "user": None, "context": {}},
            }
        ),
        lambda: EnforcementRequest("caller", "id", 1 << 64, PolicyCall()),
    ],
)
def test_request_deadline_cannot_exceed_unsigned_64_bit_range(request_factory):
    with pytest.raises(AdapterError) as raised:
        request_factory()

    assert raised.value.code == "invalid_request"


def test_adapter_revalidates_a_request_call_before_authorization(tmp_path):
    fixture = _fixture()
    adapter, audit_path = _adapter(tmp_path)
    request = EnforcementRequest.from_dict(fixture["requests"][0])
    request.call.args.append(1)

    with pytest.raises(AdapterError) as raised:
        adapter.authorize(request)

    assert raised.value.code == "invalid_request"
    assert [record["event"] for record in _records(audit_path)] == ["activation"]


def test_adapter_rejects_a_mutated_lone_surrogate_before_audit(tmp_path):
    fixture = _fixture()
    adapter, audit_path = _adapter(tmp_path)
    request = EnforcementRequest.from_dict(fixture["requests"][0])
    request.call.context["token"] = "\ud800"

    with pytest.raises(AdapterError) as raised:
        adapter.authorize(request)

    assert raised.value.code == "invalid_request"
    assert [record["event"] for record in _records(audit_path)] == ["activation"]


def test_authorization_audit_includes_metadata_but_not_raw_call_input(tmp_path):
    fixture = _fixture()
    adapter, audit_path = _adapter(tmp_path)
    request = fixture["requests"][0]
    request = {**request, "call": {**request["call"], "args": ["secret"], "context": {"token": "private"}}}

    response = adapter.authorize(EnforcementRequest.from_dict(request))

    record = _records(audit_path)[-1]
    assert response.correlation_id == "req-allow"
    assert record["caller_id"] == "host-a"
    assert record["deadline_unix_ms"] == 1001
    text = audit_path.read_text()
    assert "secret" not in text
    assert "private" not in text
    assert '"args"' not in text
    assert '"context"' not in text


def test_adapter_audit_matches_the_returned_policy_identity_and_decision(tmp_path):
    fixture = _fixture()
    adapter, audit_path = _adapter(tmp_path, now_unix_ms=fixture["now_unix_ms"])

    response = adapter.authorize(EnforcementRequest.from_dict(fixture["requests"][0]))

    record = _records(audit_path)[-1]
    assert record["event"] == "authorization"
    assert record["correlation_id"] == response.correlation_id
    assert record["policy_version"] == response.policy.version
    assert record["policy_digest"] == response.policy.digest
    assert record["decision"] == response.decision
    assert record["matched_rules"] == list(response.matched_rules)


def test_reused_correlation_id_is_audited_as_two_separate_attempts(tmp_path):
    fixture = _fixture()
    adapter, audit_path = _adapter(tmp_path)
    request = EnforcementRequest.from_dict(fixture["requests"][0])

    adapter.authorize(request)
    adapter.authorize(request)

    authorizations = [record for record in _records(audit_path) if record["event"] == "authorization"]
    assert [record["correlation_id"] for record in authorizations] == ["req-allow", "req-allow"]


def test_audit_failure_fails_closed_with_stable_adapter_error(tmp_path):
    fixture = _fixture()
    audit_path = tmp_path / "audit.jsonl"
    audit_log = AuditLog(audit_path)
    store = AuditedPolicyStore(fixture["policy"], audit_log)
    adapter = InlineEnforcementAdapter(store, now_unix_ms=lambda: fixture["now_unix_ms"])

    def fail_write(_record):
        raise AuditError("audit_write_failed", "disk unavailable")

    audit_log.append = fail_write

    with pytest.raises(AdapterError) as raised:
        adapter.authorize(EnforcementRequest.from_dict(fixture["requests"][0]))

    assert raised.value.code == "audit_write_failed"


def test_closed_adapter_maps_the_closed_audit_boundary_to_a_stable_error(tmp_path):
    fixture = _fixture()
    adapter, _ = _adapter(tmp_path, now_unix_ms=fixture["now_unix_ms"])
    adapter.close()

    with pytest.raises(AdapterError) as raised:
        adapter.authorize(EnforcementRequest.from_dict(fixture["requests"][0]))

    assert raised.value.code == "audit_write_failed"

File: tests/test_adapter.rs
use hp_guard::{
    AdapterError, AuditLog, AuditLogConfig, AuditedPolicyStore, EnforcementRequest,
    InlineEnforcementAdapter,
};
use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

fn fixture() -> Value {
    serde_json::from_str(
        &fs::read_to_string("conformance/cases/inline_adapter_v1.json").expect("read fixture"),
    )
    .expect("valid fixture")
}

fn audit_path(name: &str) -> PathBuf {
    static NEXT: AtomicUsize = AtomicUsize::new(0);
    let sequence = NEXT.fetch_add(1, Ordering::Relaxed);
    let directory = std::env::temp_dir().join(format!(
        "hp-guard-adapter-{name}-{}-{sequence}",
        std::process::id()
    ));
    fs::create_dir_all(&directory).expect("create test directory");
    directory.join("audit.jsonl")
}

fn records(path: &Path) -> Vec<Value> {
    fs::read_to_string(path)
        .expect("read audit log")
        .lines()
        .map(|line| serde_json::from_str(line).expect("valid JSON Line"))
        .collect()
}

fn adapter(policy: &str, path: &Path) -> InlineEnforcementAdapter {
    let log = AuditLog::new(path, AuditLogConfig::default()).expect("valid audit log");
    let store = AuditedPolicyStore::with_policy(policy, log).expect("activate policy");
    InlineEnforcementAdapter::with_clock(store, Arc::new(|| 1000))
}

#[test]
fn serializes_shared_fixture_responses_and_only_allows_effects() {
    let fixture = fixture();
    let path = audit_path("fixture");
    let mut adapter = adapter(fixture["policy"].as_str().expect("policy"), &path);

    let actual = fixture["requests"]
        .as_array()
        .expect("requests")
        .iter()
        .map(|value| {
            let request = EnforcementRequest::from_value(value).expect("valid request");
            serde_json::to_value(adapter.authorize(request).expect("authorized response"))
                .expect("serialize response")
        })
        .collect::<Vec<_>>();

    assert_eq!(
        actual,
        fixture["responses"].as_array().expect("responses").to_vec()
    );
    let audit_records = records(&path);
    assert_eq!(audit_records.len(), 4);
    for record in &audit_records[1..] {
        assert_eq!(record["event"], "authorization");
        assert_eq!(record["caller_id"], "host-a");
        assert_eq!(record["deadline_unix_ms"], 1001);
        assert!(record.get("args").is_none());
        assert!(record.get("context").is_none());
        assert!(!record.to_string().contains("/tmp/x"));
    }
}

#[test]
fn adapter_audit_matches_the_returned_policy_identity_and_decision() {
    let fixture = fixture();
    let path = audit_path("response-audit-identity");
    let mut adapter = adapter(fixture["policy"].as_str().expect("policy"), &path);
    let request = EnforcementRequest::from_value(&fixture["requests"][0]).expect("request");

    let response = adapter.authorize(request).expect("authorized response");

    let audit_records = records(&path);
    let record = audit_records.last().expect("authorization record");
    assert_eq!(record["event"], "authorization");
    assert_eq!(record["correlation_id"], response.correlation_id);
    assert_eq!(record["policy_version"], response.policy.version);
    assert_eq!(record["policy_digest"], response.policy.sha256);
    assert_eq!(record["decision"], serde_json::json!(response.decision));
    assert_eq!(
        record["matched_rules"],
        serde_json::json!(response.matched_rules)
    );
}

#[test]
fn rejects_expired_and_invalid_requests_before_writing_authorization() {
    let fixture = fixture();
    let path = audit_path("validation");
    let mut adapter = adapter(fixture["policy"].as_str().expect("policy"), &path);

    for case in fixture["error_cases"].as_array().expect("error cases") {
        let request = EnforcementRequest::from_value(&case["request"]);
        let error = match request {
            Ok(request) => adapter.authorize(request).expect_err("request must fail"),
            Err(error) => error,
        };
        assert_eq!(error.code(), case["error"].as_str().expect("error code"));
    }
    assert_eq!(records(&path).len(), 1, "only activation is persisted");
}

#[test]
fn reused_correlation_ids_are_audited_as_separate_attempts() {
    let fixture = fixture();
    let path = audit_path("reused-correlation");
    let mut adapter = adapter(fixture["policy"].as_str().expect("policy"), &path);
    let mut request = EnforcementRequest::from_value(&fixture["requests"][0]).expect("request");
    request.correlation_id = "retry-1".into();

    let first = adapter.authorize(request.clone()).expect("first response");
    let second = adapter.authorize(request).expect("second response");
    assert_eq!(first.correlation_id, "retry-1");
    assert_eq!(second.correlation_id, "retry-1");
    let audit_records = records(&path);
    assert_eq!(audit_records.len(), 3);
    assert_eq!(audit_records[1]["correlation_id"], "retry-1");
    assert_eq!(audit_records[2]["correlation_id"], "retry-1");
}

#[test]
fn closes_when_the_required_audit_write_fails() {
    let fixture = fixture();
    let path = audit_path("audit-failure");
    let mut adapter = adapter(fixture["policy"].as_str().expect("policy"), &path);
    fs::remove_file(&path).expect("remove audit log");
    fs::remove_dir(path.parent().expect("audit parent")).expect("remove audit directory");
    let request = EnforcementRequest::from_value(&fixture["requests"][0]).expect("request");

    assert!(matches!(
        adapter.authorize(request),
        Err(AdapterError::AuditWriteFailed)
    ));
}

#[test]
fn validates_identifier_lengths_and_complete_json_shape() {
    let fixture = fixture();
    let mut invalid = fixture["requests"][0].clone();
    invalid["caller_id"] = json!("x".repeat(257));
    assert!(matches!(
        EnforcementRequest::from_value(&invalid),
        Err(AdapterError::InvalidRequest)
    ));

    invalid = fixture["requests"][0].clone();
    invalid["extra"] = json!(true);
    assert!(matches!(
        EnforcementRequest::from_value(&invalid),
        Err(AdapterError::InvalidRequest)
    ));

    invalid = fixture["requests"][0].clone();
    invalid["call"]["context"] = json!({"attempt": 1});
    assert!(matches!(
        EnforcementRequest::from_value(&invalid),
        Err(AdapterError::InvalidRequest)
    ));
}
